source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1679 @@
+
+Example:
+
+```console
+echo "x: x + 1" > file.nix
+```
+
+```nix
+import ./file.nix 1
+```
+
+    2
+
+<details><summary>Detailed explanation</summary>
+
+The preceding shell command writes the contents `x: x + 1` to the file `file.nix` in the current directory.
+
+The above Nix expression refers to this file as `./file.nix`.
+`import ./file.nix` reads the file as a side effect and evaluates to the contained Nix expression.
+
+It is an error if the file system path does not exist.
+
+After reading the file, the Nix expression `import ./file.nix` is equivalent to the file contents:
+
+```nix
+(x: x + 1) 1
+```
+
+    2
+
+This applies the function `x: x + 1` to the argument `1`, and therefore evaluates to `3`.
+
+:::{note}
+Parentheses are required to separate function declaration from function application.
+:::
+
+</details>
+
+### `pkgs.lib`
+
+The [`nixpkgs`][nixpkgs] repository contains an attribute set called [`lib`][nixpkgs-lib], which provides a large number of useful functions.
+
+:::{note}
+The Nixpkgs manual lists all [Nixpkgs library functions][nixpkgs-functions].
+:::
+
+These functions are usually accessed through `pkgs.lib`.
+
+Example:
+
+    pkgs.lib.strings.toUpper
+
+[nixpkgs-functions]: https://nixos.org/manual/nixpkgs/stable/#sec-functions-library
+[nixpkgs-lib]: https://github.com/NixOS/nixpkgs/blob/master/lib/default.nix
+
+(side-effects)=
+## Side effects
+
+So far we have only covered what we call *pure expressions*:
+declaring data and transforming it with functions.
+
+Building software requires interaction with the outside world, called *side effects*.
+
+There is only one side effect in the Nix language that is relevant here:
+reading files from the file system as *build inputs*
+
+:::{note}
+Side effects are performed while evaluating a given impure expression.
+:::
+
+Build inputs are files that build tasks refer to in their precise description of how to derive new files.
+
+Since Nix language evaluation is otherwise pure, the only way to specify build inputs is explicitly with:
+
+- File system paths
+- Dedicated impure functions.
+
+When run, a build task will only have access to explicitly declared build inputs.
+
+:::{important}
+Purity is the key to reproducible builds.
+
+It precludes build tasks from referring to files which are not explicitly specified as build inputs.
+
+Nix supports other types of impure expressions, such as [search paths](search-path) or the constant [`builtins.currentSystem`](https://nixos.org/manual/nix/stable/language/builtin-constants.html#builtins-currentSystem).
+We do not cover those here more detail, as they do not matter for how the Nix language works in principle, and because they are discouraged for the very reason of breaking reproducibility.
+:::
+
+### Paths
+
+Whenever a file system path is rendered to a character string with [antiquotation](antiquotation), the contents of that file are copied to a special location in the file system, the *Nix store*, as a side effect.
```

Comment:
The link syntax is broken here

Reply:
If you find bugs, please leave suggestions I can merge immediately. Otherwise I'll have to dig through each comment individually to figure out what exactly was wrong and how to fix it.